sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIServerSentEvents.java

```diff
@@ -0,0 +1,77 @@
+package com.azure.ai.openai.implementation;
+
+import com.azure.core.util.serializer.JsonSerializer;
+import com.azure.core.util.serializer.JsonSerializerProviders;
+import com.azure.core.util.serializer.TypeReference;
+import reactor.core.publisher.Flux;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class OpenAIServerSentEvents<T> {
+
+    private static final JsonSerializer JSON_SERIALIZER = JsonSerializerProviders.createInstance(true);
+    private final Flux<ByteBuffer> source;
+    private final Class<T> type;
+    private AtomicReference<String> lastLine = new AtomicReference<>("");
+    private AtomicBoolean expectEmptyLine = new AtomicBoolean();
+
+    public OpenAIServerSentEvents(Flux<ByteBuffer> source, Class<T> type) {
+        this.source = source;
+        this.type = type;
+    }
+
+    public Flux<T> getEvents() {
+        return source.concatMap(byteBuffer -> {
+            try {
+                ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(byteBuffer.array());
+                BufferedReader reader = new BufferedReader(new InputStreamReader(byteArrayInputStream));
+                String currentLine = reader.readLine();
+                currentLine = lastLine.get() + currentLine;
+                List<T> values = new ArrayList<>();
+                while (currentLine != null) {
+                    if (expectEmptyLine.get() && !currentLine.isEmpty()) {
+                        return Flux.error(new UnsupportedOperationException("Multi-line data not supported"));
+                    }
+
+                    if (!expectEmptyLine.get()) {
+                        expectEmptyLine.set(true);
+                        String[] split = currentLine.split(":", 2);
```

Comment:
I believe the stream also sends a `[DONE]` message when it is in fact done. Should we check for that? I am viscerally opposed to the idea, but can't quite justify why. Is it not a standard way of finishing a stream?

Reply:
Added a check to ensure the stream ends gracefully.